### PR TITLE
feat(CategoryTheory): define classes describing the property of being Kan extended along a functor

### PR DESCRIPTION
**Pull request by dagurtomas**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Limits/KanExtension.lean [hunk 1]
```diff
@@ -197,6 +208,32 @@ theorem reflective [ι.Full] [ι.Faithful] [∀ X, HasLimitsOfShape (StructuredA
...
+/-- The right Kan extension of a functor along its restriction. -/
+abbrev self (F : L ⥤ D) [∀ x, HasLimit (Ran.diagram ι (ι ⋙ F) x)] : L ⥤ D :=
+  Ran.loc ι (ι ⋙ F)
+
+end Ran
+
+/-- A class describing the property of being right Kan extended along a functor. -/
+class IsRightKanExtendedAlong (F : L ⥤ D) [∀ x, HasLimit (Ran.diagram ι (ι ⋙ F) x)] : Prop where
+  /-- There exists an isomorphism to the right kan extension -/
+  isoSelf : Nonempty (F ≅ Ran.self ι F)
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Limits/KanExtension.lean:222]: Does an arbitrary isomorphism work? Shouldn't it be `IsIso ((Ran.adjunction D ι).unit.app F)`?

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Limits/KanExtension.lean:222]: I guess it depends on what you mean by "work". It's certainly more natural that way, but it requires the "global" Kan extension to exist. But there still is a canonical map that we can ask to be an iso. Let me mark this as WIP for now and try it out

**Comment by joelriou** [comment]
In order to formalize derived functors, I have started some PRs about (left) Kan extensions of functors (similar to @yuma-mizuno definitions for bicategories), and we already have the notion of pointwise (left) Kan extension (see https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Functor/KanExtension/Pointwise.html#CategoryTheory.Functor.LeftExtension.IsPointwiseLeftKanExtension ).

I am in the process of completely refactoring `Lan/Ran` using these notions https://github.com/leanprover-community/mathlib4/pull/10425

Then, the property you are trying to define (saying that that `F` is an extension) means that the identity, as a natural transformation `ι ⋙ F ⟶ ι ⋙ F`, is a (poinwise) right Kan extension. I think that the general API for Kan extensions should be defined first. (I plan to start completing this API first for left Kan extensions, and then dualize...)

**Comment by joelriou** [comment]
(Just to clarify, with the new general API for Kan extensions, there will be equivalences expressing that if two functors are isomorphic, then one is a Kan extension iff the other is, etc.)

**Pull request closed without merging**